lib/ts/recipe/oauth2/types.ts

```diff
@@ -0,0 +1,298 @@
+        userContext: any;
+    }): Promise<
+        | {
+              redirectUrl: string; // This will end up as a 302, encoding both error and success
+          }
+        | GeneralErrorResponse
+    >;
+
+    userInfoGET(input: {
+        oauth2AccessToken: string;
+        options: APIOptions;
+        userContext: any;
+    }): Promise<JSONObject | { status: "UNAUTHORISED"; message: string }>;
+};
+
+export type RecipeInterface = {
+    getUserInfo(input: {
+        oauth2AccessToken: string;
+
+        userContext: any;
+    }): Promise<JSONObject>;
+
+    createAuthCode(input: {
+        clientId: string;
+        responseType: ResponseType;
+        redirectUri?: string;
+        scopes: string[];
+
+        // PKCE
+        codeChallenge?: string;
+        codeChallengeMethod?: CodeChallengeMethod;
+
+        // OpenIdConnect
+        prompt?: OAuthPromptType;
+        nonce?: string;
+
+        allQueryParams: string;
+
+        session: SessionContainer;
+        userContext: any;
+    }): Promise<
+        // State is added into the url on the endpoint level
+        | {
+              status: "OK";
+              code?: string;
+              idToken?: string;
+          }
+        | { status: "CLIENT_ERROR"; error: AuthorizationErrorCode; errorDescription: string; errorURI: string } // these are redirected back to the client
+        | { status: "AUTH_ERROR"; error: string } // This gets a redirection to the auth frontend
+    >;
+
+    createTokens(
+        input:
+            | {
+                  clientId: string;
+                  grantType: "authorization_code";
+                  clientSecret?: string;
+                  code?: string;
+                  codeVerifier?: string;
+                  redirectUri: string;
+                  scope?: string[];
+
+                  allQueryParams: string;
+
+                  userContext: any;
+              }
+            | {
+                  clientId: string;
+                  grantType: "client_credentials";
+                  clientSecret: string;
+                  scope?: string[];
+
+                  userContext: any;
+              }
+            | {
+                  clientId: string;
+                  grantType: "refresh_token";
+                  clientSecret?: string;
+                  refreshToken?: string;
+                  scope?: string[];
+
+                  userContext: any;
+              }
+    ): Promise<
+        // State is added into the url on the endpoint level
+        | {
+              status: "OK";
+              tokenType: "Bearer";
+              accessToken: string;
+              refreshToken?: string;
+              idToken?: string;
+              expiresIn: number;
+          }
+        | { status: "CLIENT_ERROR"; error: TokensErrorCode } // This results in a 400 with the error code added. The status specifies that this is a client error to be consistent with the createAuthCode
+    >;
+
+    verifyAccessToken(input: {
```

Comment:
```suggestion
    verifyAccessTokenFromDatabase(input: {
```

Reply:
Just so that users know that this will query the db (as opposed to thinking that this is stateless)